File: 0112-path-sum/0112-path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type NodeWithSum struct {
    *TreeNode
    Sum int
}

func hasPathSum(root *TreeNode, targetSum int) bool {
    if root == nil {
        return false
    }
    stack := []*NodeWithSum{
        {
            TreeNode: root,
            Sum: root.Val,
        },
    }
    
    for len(stack) != 0 {
        node := stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        
        if node.Left == nil && node.Right == nil && node.Sum == targetSum{
            return true
        }
        if node.Left != nil {
            stack = append(stack, &NodeWithSum{
                TreeNode: node.Left,
                Sum: node.Sum + node.Left.Val,
            })
        }
        if node.Right != nil {
            stack = append(stack, &NodeWithSum{
                TreeNode: node.Right,
                Sum: node.Sum + node.Right.Val,
            })
        }
    }
    
    return false
}

File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if root is None:
            return False
        stack = [(root, root.val)]
        while stack:
            node, acc = stack.pop()
            if node.left is None and node.right is None:
                if acc == targetSum:
                    return True
            if node.left:
                stack.append((node.left, acc + node.left.val))
            if node.right:
                stack.append((node.right, acc + node.right.val))
        return False        
        